src/ServiceManager.php

```diff
@@ -578,55 +581,37 @@ private function createDelegatorFromName($name, ?array $options = null)
         foreach ($this->delegators[$name] as $index => $delegatorFactory) {
             $delegatorFactory = $this->delegators[$name][$index];
 
-            if ($delegatorFactory === Proxy\LazyServiceFactory::class) {
+            if ($delegatorFactory === LazyServiceFactory::class) {
                 $delegatorFactory = $this->createLazyServiceDelegatorFactory();
-            }
-
-            if (is_string($delegatorFactory) && class_exists($delegatorFactory)) {
+            } elseif (is_string($delegatorFactory) && class_exists($delegatorFactory)) {
                 $delegatorFactory = new $delegatorFactory();
             }
 
-            if (! is_callable($delegatorFactory)) {
-                if (is_string($delegatorFactory)) {
-                    throw new ServiceNotCreatedException(sprintf(
-                        'An invalid delegator factory was registered; resolved to class or function "%s"'
-                        . ' which does not exist; please provide a valid function name or class name resolving'
-                        . ' to an implementation of %s',
-                        $delegatorFactory,
-                        DelegatorFactoryInterface::class
-                    ));
-                }
-
-                throw new ServiceNotCreatedException(sprintf(
-                    'A non-callable delegator, "%s", was provided; expected a callable or instance of "%s"',
-                    is_object($delegatorFactory) ? get_class($delegatorFactory) : gettype($delegatorFactory),
-                    DelegatorFactoryInterface::class
-                ));
-            }
+            $this->assertCallableDelegatorFactory($delegatorFactory);
 
+            /**
+             * @psalm-var callable(ContainerInterface,string,callable,array<mixed>|null):object $delegatorFactory
+             */
             $this->delegators[$name][$index] = $delegatorFactory;
 
             $creationCallback = function () use ($delegatorFactory, $name, $creationCallback, $options) {
```

Comment:
I think this closure should be `static` - we're probably leaking GC (and memory, obviously) here...